packages/node/warpsync/syncing.go

```diff
@@ -0,0 +1,384 @@
+package warpsync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/celestiaorg/smt"
+	"github.com/iotaledger/goshimmer/packages/core/epoch"
+	"github.com/iotaledger/goshimmer/packages/core/tangle"
+	"github.com/iotaledger/goshimmer/packages/node/database"
+	"github.com/iotaledger/goshimmer/packages/node/p2p"
+	wp "github.com/iotaledger/goshimmer/packages/node/warpsync/warpsyncproto"
+	"github.com/iotaledger/goshimmer/plugins/epochstorage"
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/generics/lo"
+	"github.com/iotaledger/hive.go/types"
+	"golang.org/x/crypto/blake2b"
+)
+
+type epochSyncStart struct {
+	ei          epoch.Index
+	ec          epoch.EC
+	blocksCount int64
+}
+
+type epochSyncBatchBlock struct {
+	ei    epoch.Index
+	ec    epoch.EC
+	block *tangle.Block
+	peer  *peer.Peer
+}
+
+type epochSyncEnd struct {
+	ei                epoch.Index
+	ec                epoch.EC
+	stateMutationRoot epoch.MerkleRoot
+	stateRoot         epoch.MerkleRoot
+	manaRoot          epoch.MerkleRoot
+}
+
+type blockReceived struct {
+	block *tangle.Block
+	peer  *peer.Peer
+}
+
+func (m *Manager) SyncRange(ctx context.Context, startEpoch, endEpoch epoch.Index, startEC epoch.EC, ecChain map[epoch.Index]*epoch.ECRecord) error {
+	if m.syncingInProgress {
+		return fmt.Errorf("epoch syncing already in progress")
+	}
+
+	if m.isStopped {
+		return fmt.Errorf("warpsync manager is stopped")
+	}
+
+	m.syncingInProgress = true
+	defer func() { m.syncingInProgress = false }()
+
+	epochBlocks := make(map[epoch.Index][]*blockReceived)
+	tangleRoots := make(map[epoch.Index]*smt.SparseMerkleTree)
+	epochSyncEnds := make(map[epoch.Index]*epochSyncEnd)
+
+	// We do not request the start nor the ending epoch, as we know the beginning and the end of the chain.
+	for ei := endEpoch - 1; ei > startEpoch; ei-- {
+		db, _ := database.NewMemDB()
+		tangleRoots[ei] = smt.NewSparseMerkleTree(db.NewStore(), db.NewStore(), lo.PanicOnErr(blake2b.New256(nil)))
+		m.requestEpochBlocks(ei, epoch.ComputeEC(ecChain[ei]))
```

Comment:
Shouldn't this use the `concurrency` option? Also doing all in parallel seems like there might be **lots** of data coming in and stored in memory which could lead to OOM if there's lots of epochs to sync